### PR TITLE
Feature/ft232h

### PR DESCRIPTION
**Pull request by ctacke**

### CODE REVIEW

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                if (Interlocked.Decrement(ref _instanceCount) == 0)
+                {
+                    // last instance was disposed, clean house
+                    Native.Functions.Cleanup_libMPSSE();
+                }
+                _isDisposed = true;
+            }
+        }
+
+        public I2cChannel[] GetI2CChannels()
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.cs:36]: Can this follow the `II2cController` contract and implement `CreateI2cBus()`? There is an overload that takes a `busNumber`, I would think that the `channel` concept here translates to that. 

Maybe upon instantiation or lazily, whichever, we can interrogate the chip and create the appropriate `ChannelInfo`s?

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.Channel.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using static Meadow.Foundation.ICs.IOExpanders.Native;
+
+namespace Meadow.Foundation.ICs.IOExpanders
+{
+    public partial class Ft232h
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.Channel.cs:6]: maybe

`public partial class Ft232h : IDigitalInputOutputController, II2cController, ISpiController`?

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2CClockRate.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Meadow.Foundation.ICs.IOExpanders
+{
+    public partial class Ft232h
+    {
+        public enum I2CClockRate
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2CClockRate.cs:5]: Not specific to this, but I think that `II2cController` should probably have a `SupportedClockRates` or something.

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2cChannel.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            private const byte DefaultLatencyTimer = 10;
+            private const int DefaultChannelOptions = 0;
+
+            internal I2cChannel(int channelNumber, FT_DEVICE_LIST_INFO_NODE infoNode)
+                : base(channelNumber, infoNode)
+            {
+            }
+
+            public void Open(I2CClockRate clockRate = I2CClockRate.Standard)
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2cChannel.cs:18]: Maybe this is part of the `CreateI2cBus` call.

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2cChannel.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                        ClockRate = clockRate,
+                        LatencyTimer = DefaultLatencyTimer,
+                        Options = DefaultChannelOptions
+                    };
+
+                    CheckStatus(Functions.I2C_InitChannel(Handle, ref config));
+                }
+            }
+
+            protected override void CloseChannel()
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2cChannel.cs:35]: And this would be part of the `Dispose()`

**Review by bryancostanich — COMMENTED** [review]
Super stoked on this. Architecturally, I wonder if we can make it fit the existing `IIOController` pattern.

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2cChannel.cs (hunk 5 above)
Comment (ctacke) [reply Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2cChannel.cs:35]: Yes, Dispose already calls this

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.cs [hunk 6]
```diff
@@ -1,15 +1,19 @@
...
+using System.Collections.Generic;
 using System.Threading;
 
 namespace Meadow.Foundation.ICs.IOExpanders
 {
     /// <summary>
     /// Represents a DS3502 digital potentiometer
     /// </summary>
-    public partial class Ft232h : IDisposable
+    public partial class Ft232h : IDisposable, IIoDevice
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.cs:12]: Also `I2cController`, etc., yeah?

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2CClockRate.cs [hunk 7]
```diff
@@ -1,13 +1,10 @@
 ﻿namespace Meadow.Foundation.ICs.IOExpanders
 {
-    public partial class Ft232h
+    public enum I2CClockRate
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/I2CClockRate.cs:3]: `SupportedI2cFrequencies`? or similar?

**Review by bryancostanich — APPROVED** [review]
this is awesome

File: Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.cs [hunk 8]
```diff
@@ -14,6 +14,7 @@ public partial class Ft232h : IDisposable, IIoDevice
         private bool _isDisposed;
         private static int _instanceCount = 0;
         private Dictionary<int, Ft232I2cBus> _i2cBusses = new Dictionary<int, Ft232I2cBus>();
+        private Dictionary<int, Ft232SpiBus> _spiBusses = new Dictionary<int, Ft232SpiBus>();
```
Comment (bryancostanich) [inline Source/Meadow.Foundation.Peripherals/ICs.IOExpanders.Ft232h/Driver/Ft232h.cs:17]: `Buses`, not `Busses`

Might want to find/replace

**Review by adrianstevens — APPROVED** [review]

**Pull request merged**